### ISSUE TITLE
The encoder's output on code sits at the f16 ceiling (65,120 of 65,504)

### ISSUE THREAD

**Issue by MatthewLacerda2** [MatthewLacerda2/TinyRefinementModel#235]
Found with the new trajectory instrument (#225) on champion 3899391.

## Measured

Gate pre-activations in the refine loop, real (non-pad) positions, depth 8:

```
fineweb-edu   pass    preact mean    preact sd        p1        p50        p99
                 1        -6.58         7.80      -21.5       -7.7       16.4
                 2    -14803.28      7337.72   -33280.0   -14368.0    -1456.0
                 8    -15310.10      7532.53   -33952.0   -14872.0    -1468.0

codeparrot       1          inf          nan   -25184.0       -6.6    35200.0
                 2     -5351.97      9491.70   -26464.0    -5008.0    30208.0
                 8     -7008.19      9672.66   -28304.0    -6644.0    28576.0
```

`f16` maxes at 65,504. Codeparrot pass 1 **overflows to `inf`**.

### 1. The gate cannot learn any more

`sigmoid` at ±15,000 is exactly 0.0 or exactly 1.0, and its gradient there is exactly zero.
Every gate output in the model is strictly binary — verified directly: the f32 mean of `g` and
the fraction of `g > 0.5` agree to every digit, so no intermediate value exists anywhere.

`gate_bias` initializes at 0.0, i.e. `sigmoid(0) = 0.5`, a balanced gate that can be trained in
either direction. The trained model is a hard binary router that **froze** at whatever routing
it had when the pre-activations blew past the saturation point. Whatever the loop learned after
that moment, it learned with a dead gate.

This is not a small corner of the model. On prose the routing is 6 open channels of 960, the
same six at every position and every pass — so 99.375% of the refine loop's output is
multiplied by a hard zero, permanently, and no gradient flows to change it.

### 2. It is a live candidate mechanism for #229

#229: this architecture returns all-NaN logits on ~10% of code documents at 3.20B tokens, 0%
on prose, and the champion at 3.99B is clean. Measured there: an affected document is
non-finite at **every one of its 512 positions**, not a scattered few.

The gate pre-activation overflow matches on every axis:

- it is **on code**, which is where #229 fires and where prose is clean
- it reaches `inf` **already**, on the champion, at pass 1
- the gate output multiplies the whole hidden state, so one overflow contaminates the entire
  window — which is the whole-window signature #229 measured and #233 measured separately

Not proven. The cheap test is to re-run the #229 repro under `FORCE_F32_COMPUTE=1` while
logging the gate pre-activation maximum, and see whether the NaN documents are exactly the ones
whose gate overflows.

## Why the pre-activations are free to explode

`self.gate = nnx.Linear(2 * dim, dim, ...)` (`trm/model/refiner.py`), and its output goes
straight into `jax.nn.sigmoid`. There is no normalization on the gate input, no bound on its
output, and nothing in the training loop watches it. Every other path into the refine block is
RMSNorm'd; this one is not.

## Options, cheapest first

1. **Log it.** The gate pre-activation magnitude should be telemetry, not something discovered
   by hand two weeks after a run ends. It is one scalar per pass.
2. **Norm the gate input**, the way every other input to the block already is.
3. **Bound the pre-activation** (a `tanh` scaling, or a smaller gate init) so the sigmoid keeps
   a usable gradient.
4. Re-examine whether a hard binary router is what we wanted from a blend coefficient at all.
   #211 already asks whether the gate needs `2d x d` parameters; this is evidence that the
   question is worth more than a parameter count — the gate is not doing what its shape says
   it is doing.

Any change here alters what the model *is*, so it is an `ideas` item and gated behind a matched
ablation, not a fix to apply. Item 1 is `tools` and can land whenever.

Related: #225 (the instrument that found it), #229 (the NaN), #211 (the gate's parameter cost),
#233.

**Comment by MatthewLacerda2** [comment]
## Substantial correction: the gate is a symptom, not the cause

A peer session challenged the overflow arithmetic in this issue — ±15,000–35,000 fits in f16
with room to spare, so an `inf` had to be coming from somewhere else. Chasing that down moved
the root cause two layers upstream, and most of the framing above is wrong.

### What is actually oversized

```
                       max|encoder output|   max|refine block output|   max|state|
fineweb-edu                      47.7              ~30,000                 48
codeparrot                   65,120                ~30,000             56,032
                                                                 (f16 max = 65,504)
```

**The encoder output on code is 65,120 — 1,365× the prose value, and 99.4% of the way to the
f16 ceiling.** Measured identically on `best/3121279` (3.20B, 65,440) and the champion (3.99B,
65,120), so it is a stable property, not late-run drift.

The gate's own parameters are tiny: `max|w| = 0.10`, `max|b| = 0.04`. Its pre-activations reach
10⁴ purely because its input does. So:

- **"The gate learned to shut on prose and open on code" is not supported.** The gate did not
  learn a routing policy; it is a small linear map being driven into saturation by a state four
  orders of magnitude larger than its weights. Retracted.
- The binary routing is real and still worth understanding, but it is a *consequence* of the
  state magnitude, and any story about what the gate "decided" has to survive that.
- The prose/code routing difference may be partly an artifact of **which concatenated term
  dominates**: `gate(concat([z_new, z]))` sees z_new≈30,000 vs z≈48 on prose (block-dominated)
  and z_new≈30,000 vs z≈56,032 on code (state-dominated). That is a different regime, not
  necessarily a different decision.

### The gradient claim was also too strong

A saturated sigmoid kills the gradient to the **gate's own parameters**. It does not stop
gradient flowing through a channel sitting at exactly 1.0 — that is a fully open path, and the
refine block behind it kept training normally. "Everything after that trained with a dead gate"
overstated it: the *router* stopped adapting; the block did not. Correction owed to the same
peer.

I also cannot date when it saturated. Only two checkpoints survive on disk (3.20B and 3.99B)
and both are already saturated. `checkpoints_quarantine` on the HDD holds earlier steps around
the opt-step-11k collapse, which would make the onset datable.

### What this DOES establish

The overflow is real and now measured rather than inferred: on codeparrot pass 1, the f32
computation reaches **74,292.9** where f16 caps at 65,504, producing 243 non-finite values.
Not an accumulation artifact — the true value exceeds the ceiling.

And it lands the root cause on **the encoder**, which is exactly where #229 said to look:
depth 1 is enough to trigger that NaN, and z_enc is computed before any refinement. See the
comment I am posting there.

### Revised action

The actionable item is no longer about the gate. It is: **the encoder's output magnitude on
code is one rounding away from f16 overflow, and nothing measures it.** Logging
`max|z_enc|` per corpus is `tools`, cheap, and should exist regardless of what we decide to do
about the cause.

**Comment by MatthewLacerda2** [comment]
## Fully localized: the SwiGLU product in the last encoder block

Where the magnitude appears, per layer (max|activation| after embed, then each of 7 blocks):

```
fineweb-edu    0    8   12   14   17   17   22       48
codeparrot     0    7   14   21   27   31   82   65,120
                                              ↑ block 7 multiplies by ~794
```

The first six blocks behave almost identically on both corpora. **It is not rogue channels**:
898 of 960 channels exceed 1000 on code, median channel 6,768, top-8 only 4.9% of the mass —
the whole stream inflates, not a handful of dimensions.

**Block 7's weights are entirely normal.** max|w| 1.876 against 1.891 in block 6; rms 0.0352
against a smooth 0.0321→0.0347 progression across blocks 1–6. There is no weight pathology.

Inside block 7:

```
                  in    norm1   attn   norm2    block out
fineweb-edu       17      4      24      6          36
codeparrot        79      5      83      6      64,896
```

Attention is healthy on both, and `norm2` emits **the same magnitude (≈6) on both corpora**. So
the divergence is entirely in the MLP, and it is not driven by input scale:

```
block 7       norm2 in   gate_proj   silu(gate)   up_proj    product    down_proj
fineweb-edu          6         39           13         40        126           26
codeparrot           6        194          194        210     39,104       64,896

block 6 (same pattern, smaller)
fineweb-edu          5         12           11         26        163           11
codeparrot           6         25           25         51      1,211           73
```

## The mechanism

SwiGLU **multiplies** two projections of the same input. Code-shaped inputs align with
high-gain directions in both, so each factor is ~5× larger — and the product is ~310× larger,
compounded by `silu` being near-linear once the pre-activation clears the negative regime
(gate 39→194 is 5×, but silu(gate) 13→194 is 15×). `down_proj` then carries it to 64,896.

Nothing normalizes the MLP **output** before it re-enters the residual stream. The pre-norms
(`norm1`, `norm2`) bound what goes *in*; nothing bounds what comes *out*.

This is the well-documented outlier-feature / massive-activation behaviour in transformer MLPs
(the phenomenon behind LLM.int8()'s outlier handling, and Sun et al. 2024 on massive
activations). It is not exotic, and it is not specific to this architecture — what is specific
is that at f16 with no output norm, it reaches the representable ceiling.

## Why this is a next-run change, not a fix

Bounding the MLP output changes what the model *is*. The champion's weights were trained
without it and do not survive it. So this is an `ideas` item gated behind a matched ablation,
not a patch — and it lands in the next training run or not at all.

**Candidate intervention, and it is a settled technique rather than a research bet:** a
post-MLP (and post-attention) normalization on the residual branch — the "sandwich norm" /
post-norm arrangement used by Gemma 2 among others. It directly bounds the quantity that
overflows here. The matched ablation is the standard one: same seed, same data order, one knob.

## What can be done now, with no ablation

**Telemetry.** `max|z_enc|` and the per-block activation maximum should be logged during
training. The number that decides whether this model can be served in f16 at all has never been
measured during a run — this whole chain was found by hand, two weeks after the run ended.
That is `tools`, cheap, and it also subsumes items 2 and 4 inherited from #229 (a NaN-incidence
smoke, and recording activation headroom on anything we keep).

**Comment by MatthewLacerda2** [comment]
superseded-by the depth-recurrence kill (2026-09-12): the ceiling-hugging encoder output is a refiner-specific finding, and the arch-generic guard already landed — act_max is logged every step and bounded at half the f16 ceiling by instruments/invariants.py (#252), plus the out-of-vocab clamp (#233).
